networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/AttestationValidator.java

```diff
@@ -0,0 +1,212 @@
+    if (!CommitteeUtil.getSubnetId(attestation).equals(receivedOnSubnetId)) {
+      return INVALID;
+    }
+
+    // The attestation is unaggregated -- that is, it has exactly one participating validator
+    // (len([bit for bit in attestation.aggregation_bits if bit == 0b1]) == 1).
+    if (attestation.getAggregation_bits().countSetBits() != 1) {
+      return INVALID;
+    }
+
+    // The attestation is the first valid attestation received for the participating validator for
+    // the slot, attestation.data.slot.
+    if (receivedValidAttestations.contains(getValidatorAndSlot(attestation))) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidationResult singleOrAggregateAttestationChecks(final Attestation attestation) {
+    // attestation.data.slot is within the last ATTESTATION_PROPAGATION_SLOT_RANGE slots (within a
+    // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance) -- i.e. attestation.data.slot +
+    // ATTESTATION_PROPAGATION_SLOT_RANGE >= current_slot >= attestation.data.slot (a client MAY
+    // queue
+    // future attestations for processing at the appropriate slot).
+    final UnsignedLong currentTimeMillis = secondsToMillis(recentChainData.getStore().getTime());
+    if (isAfterPropagationSlotRange(currentTimeMillis, attestation)) {
+      return INVALID;
+    }
+    if (isBeforeMinimumBroadcastTime(attestation, currentTimeMillis) < 0) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    // The block being voted for (attestation.data.beacon_block_root) passes validation.
+    // It must pass validation to be in the store.
+    // If it's not in the store, it may not have been processed yet so save for future.
+    final Optional<BeaconState> maybeState =
+        recentChainData.getBlockState(attestation.getData().getBeacon_block_root());
+    if (maybeState.isEmpty()) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    final BeaconState state = maybeState.get();
+
+    // The signature of attestation is valid.
+    final IndexedAttestation indexedAttestation = get_indexed_attestation(state, attestation);
+    if (!is_valid_indexed_attestation(state, indexedAttestation)) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidatorAndSlot getValidatorAndSlot(final Attestation attestation) {
+    return new ValidatorAndSlot(
+        attestation.getData().getSlot(),
+        attestation.getData().getIndex(),
+        attestation.getAggregation_bits().streamAllSetBits().findFirst().orElseThrow());
+  }
+
+  private int isBeforeMinimumBroadcastTime(
+      final Attestation attestation, final UnsignedLong currentTimeMillis) {
+    final UnsignedLong minimumBroadcastTimeMillis =
+        minimumBroadcastTimeMillis(attestation.getData().getSlot());
+    return currentTimeMillis.compareTo(minimumBroadcastTimeMillis);
+  }
+
+  private boolean isAfterPropagationSlotRange(
+      final UnsignedLong currentTimeMillis, final Attestation aggregate) {
```

Comment:
nit: (it's an attestation regardless)
```suggestion
      final UnsignedLong currentTimeMillis, final Attestation attestation) {
```

Reply:
Yep, missed a few "aggregate" that should have been renamed when I switch to validating attestations first. Fixed.